XenAdmin/Controls/USB/UsbList.cs

```diff
@@ -0,0 +1,416 @@
+            }
+
+            foreach (DataGridViewExRow row in UsbGridView.Rows)
+            {
+                if (row is UsbRow)
+                    ((UsbRow)row).DeregisterEvents();
+            }
+        }
+
+        void UsbCollectionChanged(object sender, EventArgs e)
+        {
+            Program.Invoke(this, BuildList);
+        }
+
+        void Server_PropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            Program.Invoke(this, RefreshAllItems);
+        }
+
+        private void RefreshAllItems()
+        {
+            foreach (DataGridViewExRow row in UsbGridView.Rows)
+            {
+                if (row is UsbRow)
+                    ((UsbRow)row).UpdateDetails();
+            }
+        }
+
+        private void buttonUsage_Click(object sender, EventArgs e)
+        {
+            UsbUsageDialog dialog = new UsbUsageDialog(selectedUsb);
+            dialog.ShowDialog(Program.MainWindow);
+        }
+
+        private void buttonAttach_Click(object sender, EventArgs e)
+        {
+            if (_xenObject is VM)
+            {
+                AttachUsbDialog dialog = new AttachUsbDialog(_xenObject as VM);
+                dialog.ShowDialog(Program.MainWindow);
+            }
+        }
+
+        private void buttonDetach_Click(object sender, EventArgs e)
+        {
+
+        }
+
+        private void UsbGridView_SelectionChanged(object sender, EventArgs e)
+        {
+            selectedUsb = null;
+
+            try
+            {
+                if (UsbGridView.Rows.Count <= 0)
+                {
+                    buttonUsage.Enabled = false;
+                }
+                else
+                {
+                    HostUsbRow row = (HostUsbRow)UsbGridView.SelectedRows[0];
+                    selectedUsb = row.pusb;
+                    buttonUsage.Enabled = true;
+                }
+            }
+            catch
+            {
+                buttonUsage.Enabled = false;
+            }
+            finally
+            {
+                if (null != selectedUsb && selectedUsb.passthrough_enabled)
+                    buttonUsage.Text = Messages.USBLIST_DISABLE_PASSTHROUGH;
+                else
+                    buttonUsage.Text = Messages.USBLIST_ENABLE_PASSTHROUGH;
+            }
+        }
+
+        class UsbRow : DataGridViewExRow
+        {
+            private IXenObject _xenObject;
+
+            public UsbRow(IXenObject xo)
+            {
+                _xenObject = xo;
+                xo.PropertyChanged += Server_PropertyChanged;
+            }
+
+            public virtual void UpdateDetails()
+            {
+
+            }
+
+            public void DeregisterEvents()
+            {
+                _xenObject.PropertyChanged -= Server_PropertyChanged;
+            }
+
+            private void Server_PropertyChanged(object sender, PropertyChangedEventArgs e)

```

Comment:
Please rename this method, as containing the word Server is misleading.

Reply:
Will fix.